analyses/stackanalyses/sa.go

```diff
@@ -0,0 +1,186 @@
+		log.Fatal().Err(err).Msgf(err.Error())
+	}
+	_, err = io.Copy(fw, fd)
+	if err != nil {
+		log.Fatal().Err(err).Msgf(err.Error())
+	}
+	_ = writer.WriteField("ecosystem", fileStats.Eco)
+	_ = writer.WriteField("file_path", "/tmp/bin")
+	err = writer.Close()
+	if err != nil {
+		log.Fatal().Err(err).Msgf("Error closing Buffer Writer in SA Request.")
+	}
+	apiResponse := utils.HTTPRequestMultipart(requestData, writer, manifest)
+	body := validatePostResponse(apiResponse)
+	log.Debug().Msgf("Success: saPostRequest.")
+	return body
+}
+
+// saGetRequest performs Stack Analyses GET Request to CRDA Server.
+func saGetRequest(requestParams SARequestType, saPost SAPostResponseType, back *backoff.Backoff) SAGetResponseType {
+	log.Debug().Msgf("Executing: saGetRequest.")
+	requestData := utils.HTTPRequestType{
+		Method:          http.MethodGet,
+		Endpoint:        utils.APISA + "/" + saPost.ID,
+		ThreeScaleToken: requestParams.ThreeScaleToken,
+		Host:            requestParams.Host,
+	}
+	d := back.Duration()
+	log.Debug().Msgf("Sleeping for %s", d)
+	time.Sleep(d)
+	apiResponse := utils.HTTPRequest(requestData)
+	if apiResponse.StatusCode == http.StatusAccepted {
+		// Retry till server respond 200 or Timeout Error or Exponential Backoff limit hit.
+		log.Debug().Msgf("Retying...")
+		saGetRequest(requestParams, saPost, back)
+	}
+	body := validateGetResponse(apiResponse)
+	return body
+}
+
+// validateResponse validates API Response.
+func validatePostResponse(apiResponse *http.Response) SAPostResponseType {
+	log.Debug().Msgf("Executing validatePostResponse.")
+	var body SAPostResponseType
+	err := json.NewDecoder(apiResponse.Body).Decode(&body)
+	if apiResponse.StatusCode != http.StatusOK {
+		log.Debug().Msgf("Status from Server: %d", apiResponse.StatusCode)
+		log.Fatal().Err(err).Msgf(err.Error())
+	}
+	log.Debug().Msgf("Success validatePostResponse.")
+	return body
+}
+
+// validateGetResponse validates API Response.
+func validateGetResponse(apiResponse *http.Response) SAGetResponseType {
+	log.Debug().Msgf("Executing validateGetResponse.")
+	var body SAGetResponseType
+	err := json.NewDecoder(apiResponse.Body).Decode(&body)
+	if apiResponse.StatusCode != http.StatusOK {
+		log.Debug().Msgf("Status from Server: %d", apiResponse.StatusCode)
+		log.Fatal().Err(err).Msgf(err.Error())
+	}
+	log.Debug().Msgf("Success validateGetResponse.")
+	return body
+}
+
+// readManifest Manifest File validator and reader.
+func readManifest(shellPath string, manifestFile string) ReadManifestResponse {
+	log.Debug().Msgf("Executing readManifest.")
+	stats, err := os.Stat(manifestFile)
+	if err != nil {
+		log.Fatal().Err(err).Msgf("Error")
+	}
+	fileStats := ReadManifestResponse{
+		RawFileName: stats.Name(),
+	}
+	msg := fmt.Sprintf("Support for %s is coming soon. Thanks for your Patience. :)", fileStats.RawFileName)
+	switch fileStats.RawFileName {
+	case "requirements.txt":
+		fileStats.DepsTreePath = pypi.GeneratePylist(shellPath, manifestFile)
```

Comment:
Follow factory pattern.

There must be a generic interface which is independent of any ecosystem and each ecosystem should implement it.

e.g.
// stackanalyses.go
type StackAnalysis interface {
 func Name() string
 func IsSupportedManifestFormat(manifest path.Path) boolean
 func GeneratorDependencyTree(ctx context.Context, manifest path.Path) io.Reader
}
// factory.go

func init() {

}

register(ecosystem string, generator StackAnalysis) void {
..
}

AllRegistered() .. {
}

Reply:
@arajkumar I tried implementing this.. unable to go through it.
Actually, In Golang, In order to implement an interface, there should be the same state of "type" shared by Both packages, 
Ex: https://tour.golang.org/methods/10
`type T struct` is to be used by both, in our case `pypi` and `stackanalyses` packages. which introduces the complication of cyclic import.

Then if we bring `pypi` and `stackanalyses` under the same package, Multiple ecosystems have to implement the same Interface separately, which is kinda impossible, ie. the same Method name, different definitions under one package.

Reply:
Ack.